File: app/src/main/java/com/radionula/radionula/comments/CommentsFragment.java
package com.radionula.radionula.comments;


import android.os.Build;
import android.os.Bundle;
import android.os.Message;
import androidx.fragment.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.CookieManager;
import android.webkit.WebChromeClient;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.LinearLayout;

import com.radionula.radionula.R;

/**
 * A simple {@link Fragment} subclass.
 */
public class CommentsFragment extends Fragment {

    WebView myWebView;

    final public static String APP_KEY = "330582393726734";
    final public static String BASE_DOMAIN = "http://radionula.com";


    private WebView webView,childView =null;
    private LinearLayout parentLayout;

    public CommentsFragment() {
        // Required empty public constructor
    }

    @SuppressWarnings("deprecation")
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_comments, container, false);


        parentLayout = view.findViewById(R.id.parentLayout);


        webView = new WebView(requireActivity());
        webView.setLayoutParams(getLayoutParams());

        webView.setWebViewClient(new FaceBookClient());
        webView.setWebChromeClient(new MyChromeClient());
        webView.getSettings().setJavaScriptEnabled(true);
        webView.getSettings().setJavaScriptCanOpenWindowsAutomatically(true);
        webView.getSettings().setSupportMultipleWindows(true);

        if (Build.VERSION.SDK_INT >= 21) {
            webView.getSettings().setMixedContentMode(WebSettings.MIXED_CONTENT_ALWAYS_ALLOW);
            CookieManager.getInstance().setAcceptThirdPartyCookies(webView, true);
        }

        parentLayout.addView(webView);

        webView.loadDataWithBaseURL(BASE_DOMAIN,
                "<html>" +
                        "<head></head>" +
                        "<body>" +
                        "<div id=\"fb-root\"></div><div id=\"fb-root\"></div>" +
                        "<script>" +
                        "(function(d, s, id) {var js, fjs = d.getElementsByTagName(s)[0];if (d.getElementById(id)) return;js = d.createElement(s); js.id = id;js.src = \"http://connect.facebook.net/en_US/all.js#xfbml=1&appId=" + APP_KEY + "\";fjs.parentNode.insertBefore(js, fjs);}" +
                        "(document, 'script', 'facebook-jssdk'));" +
                        "</script>" +
                        "<div class=\"fb-comments\" data-href=\"" + BASE_DOMAIN + "\" data-width=\"470\" data-order-by=\"reverse_time\" data-colorscheme=\"light\"></div> " +
                        " " +
                        " </body>" +
                        "</html>", "text/html", null, null);

        return view;
    }

    public boolean webViewGoBack() {
        if(childView != null && parentLayout.getChildCount()==2){
            childView.stopLoading();
            parentLayout.removeViewAt(parentLayout.getChildCount()-1);
            if(webView.getVisibility() == View.GONE)
                webView.setVisibility(View.VISIBLE);
            return true;
        }else{
            return false;
        }
    }


    private LinearLayout.LayoutParams getLayoutParams(){
        return new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT,LinearLayout.LayoutParams.MATCH_PARENT);
    }


    final class MyChromeClient extends WebChromeClient{
        @Override
        public boolean onCreateWindow(WebView view, boolean dialog,
                                      boolean userGesture, Message resultMsg) {
            childView = new WebView(getActivity());
            childView.getSettings().setJavaScriptEnabled(true);
            childView.setWebViewClient(new FaceBookClient());
            childView.setWebChromeClient(this);
            childView.setLayoutParams(new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT,LinearLayout.LayoutParams.MATCH_PARENT));


            parentLayout.addView(childView);


            childView.requestFocus();
            webView.setVisibility(View.GONE);

          /*I think this is the main part which handles all the log in session*/
            WebView.WebViewTransport transport =(WebView.WebViewTransport)resultMsg.obj;
            transport.setWebView(childView);
            resultMsg.sendToTarget();
            return true;
        }


        @Override
        public void onProgressChanged(WebView view, int newProgress) {
            getActivity().setProgress(newProgress * 100);
        }

        @Override
        public void onCloseWindow(WebView window) {
            parentLayout.removeViewAt(parentLayout.getChildCount()-1);
            childView =null;
            webView.setVisibility(View.VISIBLE);
            webView.requestFocus();
        }
    }

    private static class FaceBookClient extends WebViewClient{
        @Override
        public boolean shouldOverrideUrlLoading(WebView view, String url) {
            Log.i("REQUEST URL", url);
            return false;
        }
    }


}

File: app/src/main/java/com/radionula/radionula/comments/CommentsFragment.kt
package com.radionula.radionula.comments

import android.annotation.SuppressLint
import android.content.ActivityNotFoundException
import android.content.Intent
import android.graphics.Color
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.webkit.CookieManager
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.fragment.app.Fragment
import com.radionula.radionula.databinding.FragmentCommentsBinding
import com.radionula.radionula.radio.ChannelPresenter
import org.koin.android.ext.android.inject

/**
 * Remark42 comments for the channel that is currently tuned in.
 *
 * radionula.com dropped the Facebook comments plugin this screen used to embed -
 * the Remark42 instance only offers telegram and anonymous login - so this loads
 * the same widget the website does, with the same host, site id and thread URL,
 * to keep the app and the site on one conversation.
 */
class CommentsFragment : Fragment() {

    private val channelPresenter: ChannelPresenter by inject()

    private var binding: FragmentCommentsBinding? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View = FragmentCommentsBinding.inflate(inflater, container, false)
        .also { binding = it }
        .root

    @SuppressLint("SetJavaScriptEnabled")
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val webView = binding?.commentsWebView ?: return

        // Remark42 is a JavaScript widget and keeps its preferences in localStorage.
        webView.settings.javaScriptEnabled = true
        webView.settings.domStorageEnabled = true
        // The comments render inside an iframe served from the Remark42 host, so
        // its login cookie is third-party as far as the page is concerned.
        CookieManager.getInstance().setAcceptThirdPartyCookies(webView, true)
        // Leaving popups off means target=_blank links load here instead of being
        // swallowed, and shouldOverrideUrlLoading can hand them to the browser.
        webView.settings.setSupportMultipleWindows(false)
        webView.setBackgroundColor(Color.TRANSPARENT)
        webView.webViewClient = ExternalLinkClient()

        webView.loadDataWithBaseURL(SITE_URL, embedHtml(), "text/html", "utf-8", null)
    }

    override fun onDestroyView() {
        binding?.commentsWebView?.let { webView ->
            webView.stopLoading()
            (webView.parent as? ViewGroup)?.removeView(webView)
            webView.destroy()
        }
        binding = null
        super.onDestroyView()
    }

    private fun embedHtml(): String {
        val channel = channelPresenter.currentChannel
        return """
            <!doctype html>
            <html>
            <head>
              <meta name="viewport" content="width=device-width, initial-scale=1">
              <style>html, body { margin: 0; padding: 8px; background: transparent; }</style>
            </head>
            <body>
              <div id="remark42"></div>
              <script>
                var remark_config = {
                  host: "$REMARK42_HOST",
                  site_id: "$REMARK42_SITE_ID",
                  url: "${channel.commentsUrl}",
                  page_title: "Radio NULA - ${channel.displayName}",
                  theme: "dark",
                  locale: "en",
                  components: ["embed"],
                  max_shown_comments: 10
                };
              </script>
              <script src="$REMARK42_HOST/web/embed.js"></script>
            </body>
            </html>
        """.trimIndent()
    }

    private inner class ExternalLinkClient : WebViewClient() {
        override fun shouldOverrideUrlLoading(view: WebView, request: WebResourceRequest): Boolean {
            val host = request.url.host ?: return false
            if (host == SITE_HOST || host.endsWith(".$SITE_HOST")) return false

            // Telegram login, and anything else off-site, belongs in a real browser.
            return openExternally(request.url)
        }
    }

    private fun openExternally(uri: Uri): Boolean = try {
        startActivity(Intent(Intent.ACTION_VIEW, uri))
        true
    } catch (e: ActivityNotFoundException) {
        false
    }

    private companion object {
        const val SITE_HOST = "radionula.com"
        const val SITE_URL = "https://radionula.com/"
        const val REMARK42_HOST = "https://comments.radionula.com"
        const val REMARK42_SITE_ID = "radionula-prod"
    }
}
